### PR TITLE
Added mdadm support for provision

### PR DESCRIPTION
**Pull request by macdems**
This patch does two things:

1) It fixes some bugs I found in [provision/initramfs/capabilities/setup-filesystems/20-filesystems](/warewulf/warewulf3/tree/development/provision/initramfs/capabilities/setup-filesystems/20-filesystems).

2) It adds support for `mdadm` based RAID arrays. The sample syntax to create RAID0 is:

   ```
   # Parted specific commands
   select /dev/sda
   mklabel gpt
   mkpart primary 0% 100%
   set 1 raid on

   select /dev/sdb
   mklabel gpt
   mkpart primary 0% 100%
   set 1 raid on

   # Create RAID0 on /dev/md0: raidX NUMBER PARTITIONS...
   raid0 0 sda1 sdb1

   # the above command also selects /dev/md, so filesystems can be created

   # Create filesystem on /dev/md0
   mkfs 0 ext2 -L tmp
   ```

   Supported raid levels are 0, 1, 4, 5, 6, 10

### CODE REVIEW

File: provision/3rd_party/GPL/Makefile.am
Comment (bensallen) [inline provision/3rd_party/GPL/Makefile.am]: We'll need to attempt to get a maintainer (appears this is Jes Sorensen) of mdadm to sign-off that we're not a GPL derivative work. If we can find Jes' contact info I'll email him.

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (bensallen) [inline provision/initramfs/capabilities/setup-filesystems/20-filesystems]: We recently added this logic in to deal with disks not being ready. What problems did you run into with it?

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (bensallen) [inline provision/initramfs/capabilities/setup-filesystems/20-filesystems:194]: Assuming that `--force` isn't sufficient to get mdadm to not prompt?

Double quote ${DEVICES} to avoid word splitting.

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (bensallen) [inline provision/initramfs/capabilities/setup-filesystems/20-filesystems]: Nit, update comment to "...before continuing with raid" or similar

**Review by bensallen — CHANGES_REQUESTED** [review]
Hey Maciej, this is awesome! Raid support is a long standing request for the project. Couple of questions and requests.

File: provision/3rd_party/GPL/Makefile.am
Comment (macdems) [reply provision/3rd_party/GPL/Makefile.am]: Yes please, could you contact him on behalf of warewulf?

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (macdems) [reply provision/initramfs/capabilities/setup-filesystems/20-filesystems:194]: "--force" isn't sufficient. I don't like it, but `yes` seems the only way...

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (macdems) [reply provision/initramfs/capabilities/setup-filesystems/20-filesystems]: I originaly worked on master base, so I did not notice this change. I will look into it.

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (macdems) [reply provision/initramfs/capabilities/setup-filesystems/20-filesystems:194]: And I don't think double-quoting `${DEVICES}` is right. We want it bo be multiple parameters, not just one with spaces in it.

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (macdems) [reply provision/initramfs/capabilities/setup-filesystems/20-filesystems]: ok

File: provision/3rd_party/GPL/Makefile.am
Comment (bensallen) [reply provision/3rd_party/GPL/Makefile.am]: Thanks. I've sent him an email. Can you please pull his email addresses from your comment so we don't inadvertently cause any spam to head his way?

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (bensallen) [reply provision/initramfs/capabilities/setup-filesystems/20-filesystems:194]: > "--force" isn't sufficient. I don't like it, but `yes` seems the only way...

Ok. Please update the comment above stating what this is a workaround for.

> And I don't think double-quoting `${DEVICES}` is right. We want it bo be multiple parameters, not just one with spaces in it.

Yep you're correct.

File: provision/initramfs/capabilities/setup-filesystems/20-filesystems
Comment (macdems) [reply provision/initramfs/capabilities/setup-filesystems/20-filesystems]: I have moved it to separate function. This way, we can use it to wait for partition in case `udevadm` is not available.

**Comment by bensallen** [comment]
@macdems Unfortunately I've been unable to contact the maintainer of mdadm. I'm thinking we should pull the mdadm source out of the PR, and rely on `--with-local-mdadm`. I suggest we leave the Makefile mdadm target in place in case we want to go back to it, but remove mdadm from the rootfs makefile target so its not automatically built.

**Comment by macdems** [comment]
Sorry, I am unable to follow the changes in the Warewulf. At the moment I submitted the PR this worked. Now the master changed too much and I cannot merge it any more.

Do I understand correctly that `--with-local-*` disappeared?

**Comment by bensallen** [comment]
Merged in #296. Thanks again @macdems. Sorry it took me so long to get around to handling this one.

**Comment by macdems** [comment]
@bensallen thank you for merging. I'm glad, I could contribute :)

**Pull request closed without merging**
